src/main/java/emu/grasscutter/game/player/PlayerProgressManager.java

```diff
@@ -313,4 +315,32 @@ public void addItemObtainedHistory(int id, int count) {
         player.save();
         player.getQuestManager().queueEvent(QuestCond.QUEST_COND_HISTORY_GOT_ANY_ITEM, id, newCount);
     }
+
+    /******************************************************************************************************************
+     ******************************************************************************************************************
+     * SCENETAGS
+     ******************************************************************************************************************
+     *****************************************************************************************************************/
+    public void addSceneTag(int sceneId, int sceneTagId) {
+        // Sanity check
+        if (player.getSceneTags().get(sceneId) == null) {
```

Comment:
Only a small thing, but you can also do something like this to add the hashset when its not there yet:
```java
player.getSceneTags().computeIfAbsent(sceneId, (key) -> new HashSet<>())
    .add(sceneTagID)
```

Reply:
Ah of course, I was trying to recall the better inline method. That's what I get for making this right before bed.